demos/device/src/room/containers/RoomProvider/index.tsx

```diff
@@ -0,0 +1,142 @@
+import React, { useEffect, useReducer, useRef } from 'react';
+import { RouteComponentProps, withRouter } from 'react-router-dom';
+
+import MeetingManager from '../../MeetingManager';
+import routes from '../../../routes';
+import { initialState, reducer, Type as actionType } from './reducer';
+import { MessageHandler, DeviceMessage } from '../../../shim/types';
+import { Type as messageType } from '../../../controller/containers/ControllerProvider/reducer';
+import { ScreenObserver, ScreenMessageDetail } from '../../../../../../build';
+
+const screenViewDiv = () => document.getElementById('shared-content-view') as HTMLDivElement;
+const nameplateDiv = () =>
+  document.getElementById('share-content-view-nameplate') as HTMLDivElement;
+
+const sendMessage = async (msg: DeviceMessage): Promise<void> => {
+  if (!window.deviceEnvironment) return;
+  console.log(`Sending message to controller ${msg.type}`);
+
+  const env = await window.deviceEnvironment;
+  env.sendMessage(msg);
+};
+
+interface RoomProviderProps extends RouteComponentProps {}
+
+const RoomProvider: React.FC<RoomProviderProps> = ({ history }) => {
```

Comment:
You can write this logic in more React way. In the edu demo, we let each React component or container subscribe to particular Chime SDK observers and updating UIs.

For example, you can listen to `ScreenObserver` in `<VideoManager>` and pass down props to `<VideoGrid>` to re-render UIs. Or you can add one more manager `<ScreenViewManager>` for the screen view part.

Reply:
Refactored!